### ISSUE TITLE
GameGuru MAX - Bug - Light Leaking Through Models

### ISSUE THREAD

**Issue by MonkeyFrogStudio** [TheGameCreators/GameGuruRepo#4188]
I'm seeing some pretty bad light leaking through solid models from time time. Here's a small example:

![leak](https://user-images.githubusercontent.com/36735507/220977497-902dc99e-c0bf-4d90-8eb4-bb150c021edb.png)

That's one single model with all faces connected to each other. Here's the model in Blender:

![leak2](https://user-images.githubusercontent.com/36735507/220977605-44580f72-4f2c-418a-af52-c980741744f2.png)

The steps are a separate model, but the walls and ceiling are all one model, so there should be no light leaking through the cracks here.

**Comment by synchromesh62** [comment]
Ye i get the same thing. Its a bit of a GG classic trait that never really got sorted.
Must be a way to get this fixed.

**Comment by MonkeyFrogStudio** [comment]
A few more images to showcase this:

_This one shows the light leaking under a door and the walls near it:_
![leak](https://user-images.githubusercontent.com/36735507/221086974-48c2047d-f58d-447d-ba2d-14220c720624.png)

_However, here's the other side:_
![leak2](https://user-images.githubusercontent.com/36735507/221086999-4c40c023-64f7-46f6-b3b9-9b7952147c2b.png)

As you can see, the wall is thick (it has polygonal faces all around ... top, bottom, back, front, sides). In fact, the wall/door combo intersects with the sidewalk, too. So, how is light getting through there?

**Comment by Granada1** [comment]
Could you put the model up , I could try and look at it for you !

**Comment by gghermm** [comment]
Hi,Monkey Frog,Granada1,This is with all Max buildings, and walls, so.at Warehouse light blue,Is the strip not only on the floor, but also in the wall. So 3 stripes, which are also felt 5 cm wide.And always on the side facing the sun.No matter how you turn it.And even if you build cellars.At night all stripes are gone.That was from the beginning.
![Buildings](https://user-images.githubusercontent.com/125870008/221137834-00ad5021-362f-4658-85de-047a630c6392.png)

**Comment by MonkeyFrogStudio** [comment]
@Granada1 I cannot share the model at this time. However, I don't think it's the model. More to follow in next post. ;)

**Comment by MonkeyFrogStudio** [comment]
Upon further investigation, I don't think it's light leaking at all, but how shadows are rendered based on distance from the camera/player. I will try to show a few examples to illustrate what I mean.

_This image shows more of the "light leaking" in other areas":_
![leak1](https://user-images.githubusercontent.com/36735507/221216335-ef822237-deb8-4e05-9222-e0f44fbf1295.png)

_Now look at this door/wall piece (two models) when viewed close-up:_
![leak2](https://user-images.githubusercontent.com/36735507/221216610-fc73befd-0711-4608-ac49-c21d15644971.png)

No "light leaking". 

_Now let's back the camera away:_
![leak3](https://user-images.githubusercontent.com/36735507/221216759-0e60bb54-6ce3-40b4-85e0-f14dcae59ef3.png)

As you move the camera back, you can see the shadows shift and change as if they are being swapped out due to distance. The result seems to be a separation from the base of the objects (where the object meet or intersect) causing what looks like light leaking in.

_Here's another example with a different model:_
![leak4](https://user-images.githubusercontent.com/36735507/221217148-244cabcc-df7d-45d6-9782-4175c54ffeab.png)

Notice where the arrows are pointing? There appears to be "light leaking" despite the fact that the stair model is solid and it also intersects the sidewalk objects, which are also solid.

Now let's move the camera in a bit:
![leak5](https://user-images.githubusercontent.com/36735507/221217367-ddaca917-3491-4cf3-9fce-63b96b939c45.png)

The shadows have shifted a bit. As a result, one of the "light leaks" has completely disappeared, but the other still remains (red arrow).

_Now, let's move in a bit closer:_
![leak6](https://user-images.githubusercontent.com/36735507/221217572-24297d36-2571-4555-8341-098e1378f8d9.png)

The "light leaking" has completely disappeared and all's good again. 

If these were videos (and I may take some video of this later today), you could see the shadows shifting as you move closer to and further from the objects. What this does, though, is create these "light leaks" on distant objects that disappear as you get closer to the object. It can be quite visible and looks pretty bad because it looks like light is leaking through the geometry. But it appears to be how MAX is working with shadows based on distance from the camera.

**Comment by Granada1** [comment]
I think you might have found the answer 👍

**Comment by MonkeyFrogStudio** [comment]
Here's a video where you can plainly see the shadows change as the camera moves:

https://user-images.githubusercontent.com/36735507/221268116-11752e8b-ec13-43b1-845f-b4ee9d79af9b.mp4

You can see the "light leaking" disappear when the camera gets closer and the shadows change from a softer shadow that doesn't quite meet the edge or intersect of the objects to a harder shadow that does.

**Comment by Kravenwolf** [comment]
@MonkeyFrogStudio , I've just encountered this same issue, and increasing the shadow resolution under the Shadows tab (1024 did the trick for me) seems to have fixed this problem from all distances, if you wanted to give that a go to confirm (I'm running an older build so it may not still be the case).


![light_bleed](https://user-images.githubusercontent.com/103399301/231076326-f1df0320-7d1b-4b9b-9cd4-1b7d345a5884.jpg)

**Comment by MonkeyFrogStudio** [comment]
@Kravenwolf Thanks for this. I appreciate it.

In my example, it was the sun creating the issue and I had to have shadows cranked up to 2048 to make it go away, but it did go away (in the latest experimental build, at least). However, that's still an issue (in my mind) since it would mean I would need to use the highest setting for shadows, potentially slowing down the game for some. 

In any case, I still think we should be able to use lower shadow resolutions if we like without there being gaps that look like light leaking. Worse looing shadows? Yes. Of course. But obvious gaps? No. ;)

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio I agree this is more bug than feature, but moving to LOW as we will likely get a new shadow system down the road that resolves this. For now, even using the shader resolution and camera far distance setting tricks do not always work:
![image](https://github.com/TheGameCreators/GameGuruRepo/assets/35840583/97c36d91-2cad-473a-afb2-91a2b567051a)

The only crude solution that is guarenteed is a secondary ceiling hull placed higher than the interior space polygons, so you get the fix looking like this, even with small shadow res and max camera distance:
![image](https://github.com/TheGameCreators/GameGuruRepo/assets/35840583/cdda3eee-8206-42be-ae7f-96f4a17cf8ea)

The fix looks like this (exaggerated but you get the idea of the 'workaround' solution):
![image](https://github.com/TheGameCreators/GameGuruRepo/assets/35840583/490c7c56-888a-4b7d-8791-567a551148d4)

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Made a small tweak (in the next EXP build) to the shadow cascade thresholds that might improve the transitions in the shadow resolutions.  Not entirely happy with the current system, and still feel a whole new 'shadow system' is required for no-nonsense lighting but for the present DX11 real-time lighting engine, it should be okay with some careful (and opportunistic) level design to hide the occasional seams.  Will retitle this issue as a feature request to drop this four-level cascade render approach and opt for something a little more dynamic and if possible pre-computed.

**Comment by imothep85** [comment]
When this bug is going to be solved ???
